interception-vimproved.cpp

```diff
@@ -1,427 +1,569 @@
-  space->addMapping(KEY_K, KEY_UP);
-  space->addMapping(KEY_L, KEY_RIGHT);
-
-  // vim above home row
-  space->addMapping(KEY_Y, KEY_HOME);
-  space->addMapping(KEY_U, KEY_PAGEDOWN);
-  space->addMapping(KEY_I, KEY_PAGEUP);
-  space->addMapping(KEY_O, KEY_END);
-
-  // number row to F keys
-  space->addMapping(KEY_1, KEY_F1);
-  space->addMapping(KEY_2, KEY_F2);
-  space->addMapping(KEY_3, KEY_F3);
-  space->addMapping(KEY_4, KEY_F4);
-  space->addMapping(KEY_5, KEY_F5);
-  space->addMapping(KEY_6, KEY_F6);
-  space->addMapping(KEY_7, KEY_F7);
-  space->addMapping(KEY_8, KEY_F8);
-  space->addMapping(KEY_9, KEY_F9);
-  space->addMapping(KEY_0, KEY_F10);
-  space->addMapping(KEY_MINUS, KEY_F11);
-  space->addMapping(KEY_EQUAL, KEY_F12);
-
-  // xf86 audio
-  space->addMapping(KEY_M, KEY_MUTE);
-  space->addMapping(KEY_COMMA, KEY_VOLUMEDOWN);
-  space->addMapping(KEY_DOT, KEY_VOLUMEUP);
-
-  // mouse navigation
-  space->addMapping(BTN_LEFT, BTN_BACK);
-  space->addMapping(BTN_RIGHT, BTN_FORWARD);
-
-  // @FIXME: this is not working, even though `wev` says keycode 99 is Print
-  // PrtSc -> Context Menu
-  space->addMapping(KEY_SYSRQ, KEY_CONTEXT_MENU);
+  auto space = new InterceptedKeyLayer(KEY_SPACE, KEY_SPACE);
+  space->add_mappings(mappings);
 
   // tap caps for esc, hold for ctrl
-  InterceptedKeyModifier *caps =
-      new InterceptedKeyModifier(KEY_CAPSLOCK, KEY_ESC, KEY_LEFTCTRL);
+  auto caps = new InterceptedKeyModifier(KEY_CAPSLOCK, KEY_ESC, KEY_LEFTCTRL);
 
   // tap enter for enter, hold for ctrl
-  InterceptedKeyModifier *enter =
-      new InterceptedKeyModifier(KEY_ENTER, KEY_ENTER, KEY_RIGHTCTRL);
+  auto enter = new InterceptedKeyModifier(KEY_ENTER, KEY_ENTER, KEY_RIGHTCTRL);
 
-  // @NOTE: modifier keys must go first because layerKey.processInterceptedHeld
+  // NOTE: modifier keys must go first because layerKey.processInterceptedHeld
   // emits mapped key as soon as the for loop calls layerKey.process..
   // if that process is run before modifierKey.process, the modifier key will
   // not be emitted
-  vector<InterceptedKey *> *interceptedKeys = new vector<InterceptedKey *>();
-  interceptedKeys->push_back(caps);
-  interceptedKeys->push_back(enter);
-  interceptedKeys->push_back(space);
+  auto interceptedKeys = std::vector<InterceptedKey*>{caps, enter, space};
   return interceptedKeys;
 }
 
-int main() {
-  auto interceptedKeys = initInterceptedKeys();
-
-  setbuf(stdin, NULL), setbuf(stdout, NULL);
-
-  /* event *input = (event*) malloc(input_event_struct_size); */
-  Event *input = new Event();
-  while (readEvent(input)) {
-    if (input->type == EV_MSC && input->code == MSC_SCAN)
-      continue;
-
-    if (input->type != EV_KEY) {
-      writeEvent(input);
-      continue;
+auto read_mappings_or_default(int argc, char** argv) -> std::vector<Mapping> {
+  if (argc < 2) return DEFAULT_MAPPINGS;
+  try {
+    auto file = std::ifstream(argv[1]);
+    auto mappings = std::vector<Mapping>();
```

Comment:
not sure about the exact DSL, but as you implemented something like:

```
key_from1 key_to1
key_from2 key_to2
```

that only refers to which keys to be remapped when space key is held, while the "config" which key activates the layer (in other words - acts as a layer modifier key) is hardcoded as `space` in the code..

my idea is to make that configurable as well, and the first thing that came to my mind is TOML-like config file that adds another layer of abstraction to the config file:

```
[layer space space]
key_from1 key_to1
key_from2 key_to2
```

where `[]` is a header which carries info:
* "layer" means type of the mapping
* first "space" means modifier key, ie. that the config found in this header will apply "when space key is held"
* second "space" means "when the modifier key is released without chording, emit space"

similarly, we could have:
```
[modifier enter enter rightctrl]
# empty
[modifier caps esc leftctrl]
# empty
```
that would capture the rest of the hardcoded config..


now that i write it all down i see that the DSL syntax is not ideal, and it's also not a valid TOML, so not the best implementation for this use case, but something else can be used instead, for instance YAML:
```yaml
---
modifiers:
  - name: enter-to-rctrl
    intercepted: enter
    tapped: enter
    held: rightctrl
  - name: caps-to-lctrl
    intercepted: caps
    tapped: esc
    held: leftctrl
layers:
  - name: space-cadet
    intercepted: space
    tapped: space
    mappings:
      key_e: key_esc
      key_d: key_delete
      ...
```

or whatever is easy for humans to read and to impl in the code..

my first suggestion was based on introducing minimal change to the config that you already implemented, but for full flexibility and _forward compatibility_ it might be actually better to have a more structured config like the yaml from the given example.. any opinions?